frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,472 @@
+		#[pallet::weight(0)]
+		pub fn renew(
+			origin: OriginFor<T>,
+			name_hash: NameHash,
+			expiry: T::BlockNumber,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			Self::do_renew(sender, name_hash, expiry)?;
+			Ok(())
+		}
+
+		/// Deregister a registered name.
+		///
+		/// If the registration is still valid, only the owner of the name can make this call.
+		///
+		/// If the registration is expired, then anyone can call this function to make the name
+		/// available.
+		#[pallet::weight(0)]
+		pub fn deregister(origin: OriginFor<T>, name_hash: NameHash) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let registration =
+				Registrations::<T>::get(name_hash).ok_or(Error::<T>::RegistrationNotFound)?;
+			// If the registration is expired, anyone can trigger deregister.
+			if !Self::is_expired(&registration) {
+				ensure!(Self::is_owner(&registration, &sender), Error::<T>::NotOwner);
+			}
+			Self::do_deregister(name_hash);
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn set_subnode_record(
+			origin: OriginFor<T>,
+			parent_hash: NameHash,
+			label: Vec<u8>,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			Self::do_set_subnode_record(sender, parent_hash, &label)?;
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn deregister_subnode(
+			origin: OriginFor<T>,
+			parent_hash: NameHash,
+			label_hash: NameHash,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let subnode_hash = Self::subnode_hash(parent_hash, label_hash);
+			let subnode_registration = Self::get_registration(subnode_hash)?;
+			// The owner isn't calling, we check that the parent registration doesn't exist, which
+			// mean this subnode is still valid.
+			if !Self::is_owner(&subnode_registration, &sender) {
+				ensure!(
+					Self::get_registration(parent_hash).is_err(),
+					Error::<T>::RegistrationNotExpired
+				);
+			}
+			Self::do_deregister(subnode_hash);
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn set_subnode_owner(
+			origin: OriginFor<T>,
+			parent_hash: NameHash,
+			label_hash: NameHash,
+			new_owner: T::AccountId,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			Self::do_set_subnode_owner(sender, parent_hash, label_hash, new_owner)?;
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn set_address(
+			origin: OriginFor<T>,
+			name_hash: NameHash,
+			address: T::AccountId,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let registration =
+				Registrations::<T>::get(name_hash).ok_or(Error::<T>::RegistrationNotFound)?;
+			ensure!(Self::is_controller(&registration, &sender), Error::<T>::NotController);
+			T::NameServiceResolver::set_address(name_hash, address, sender)?;
+			Ok(())
+		}
+
+		/// Register the raw name for a given name hash. This can be used as a reverse lookup for
+		/// front-ends.
+		///
+		/// This is a permissionless function that anyone can call who is willing to place a deposit
```

Comment:
Is it correct to talk about deposits here? The `NameServiceResolver` trait does not really enforce it. This seems a (recommended) implementation detail.